h2o-jetty-8/src/main/java/ai/h2o/jetty8/LoginHandler.java

```diff
@@ -0,0 +1,76 @@
+package ai.h2o.jetty8;
+
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.server.handler.HandlerWrapper;
+import water.api.RequestServer;
+import water.server.ServletUtils;
+
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+
+class LoginHandler extends HandlerWrapper {
+
+  private String _loginTarget;
+  private String _errorTarget;
+
+  LoginHandler(String loginTarget, String errorTarget) {
+    _loginTarget = loginTarget;
+    _errorTarget = errorTarget;
+  }
+
+  @Override
+  public void handle(String target, Request baseRequest, HttpServletRequest request, HttpServletResponse response)
+      throws IOException, ServletException {
+    if (isLoginTarget(target)) {
+      if (isPageRequest(request)) {
+        sendLoginForm(request, response);
+      } else {
+        ServletUtils.sendResponseError(response, HttpServletResponse.SC_UNAUTHORIZED, "Access denied. Please login.");
+      }
+      baseRequest.setHandled(true);
+    } else {
+      // not for us, invoke wrapped handler
+      super.handle(target, baseRequest, request, response);
+    }
+  }
+
+  private void sendLoginForm(HttpServletRequest request, HttpServletResponse response) {
+    final String uri = ServletUtils.getDecodedUri(request);
+    try {
+      byte[] bytes;
+      try (InputStream resource = water.init.JarHash.getResource2("/login.html")) {
+        if (resource == null) {
+          throw new IllegalStateException("Login form not found");
+        }
+        final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        water.util.FileUtils.copyStream(resource, baos, 2048);
+        bytes = baos.toByteArray();
+      }
+      response.setContentType(RequestServer.MIME_HTML);
+      response.setContentLength(bytes.length);
+      ServletUtils.setResponseStatus(response, HttpServletResponse.SC_OK);
+      final OutputStream os = response.getOutputStream();
+      water.util.FileUtils.copyStream(new ByteArrayInputStream(bytes), os, 2048);
+      // TODO: this whole method can be replaced with just:
+      // org.apache.commons.io.IOUtils.copy( water.init.JarHash.getResource2("/login.html"), os);
```

Comment:
water.init.JarHash.getResource2 is caching the resources in memory

Reply:
I see, but: is showing login page so performance-critical that we want to read it into bytearray, cache this, and then copy this bytearray into response? This brings a lot of code (compared to single line) that needs to be maintained, and the only value is, that the user will see login page few microseconds earlier... My point was, the less code to maintain, the better?